Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathFinder.php

```diff
@@ -80,6 +81,7 @@ public function getByIdAndDimensionSpacePointHash(
         if ($this->cacheEnabled && isset($this->getByIdAndDimensionSpacePointHashCache[$cacheKey])) {
             return $this->getByIdAndDimensionSpacePointHashCache[$cacheKey];
         }
+        // todo AND removed = 0 -> we should not allow building uris to removed nodes!
```

Comment:
This has to be discussed!

Currently matching and resolving is not behaving the same for disabled nodes:

As shown by this test:

```
Then No node should match URL "/david-nodenborough/earl-document/leaf"
And The node "leaf-mc-node" in content stream "cs-identifier" and dimension "{}" should resolve to URL "/david-nodenborough/earl-document/leaf"
```

so matching is expected to work always, but i think this is wrong! And i dont want to introduce this for soft removed nodes.

Reply:
it would be utterly broken if we introduce it for soft deleted nodes. it must free the url.

Reply:
Oh look:

> Note: This will not exclude *disabled* nodes in order to allow the calling side to make a distinction (e.g. in order to display a custom error)

This is from the docblock at `\Neos\Neos\FrontendRouting\Projection\DocumentUriPathFinder::getByIdAndDimensionSpacePointHash`

Reply:
Which for example the shortcut resolver uses to throw a custom error about the target being disabled.

I guess we need split behavior where disabled does resolve and removed doesnt'

Reply:
or we just dont do that for now as this introduces just additional complexity. There are a lot of cases where disabled nodes cannot be worked with like when using back references or references but the nodes are hidden and thus wont be displayed in fusion live rendering. My point is, its fairly obvious that linking should not work and editors who want to find out why not will not have access to the stack trace and have to find out either way by looking at the tree, and why make it complex just to make it easy for developers.
I think we should have a clean api to the outside here.

Reply:
okay found it the resolving (not matching!) of disabled nodes was done via https://github.com/neos/neos-development-collection/pull/4363 to make the redirect handler work.

So we sadly cannot remove the complexity to return disabled nodes and have to split decisions for soft removed nodes.